accepted/better-obsoletion/better-obsoletion.md

```diff
@@ -0,0 +1,345 @@
+support the desired scenarios. Alternatives are listed afterwards.
+
+### API Changes
+
+We'd add the following properties to `ObsoleteAttribute`:
+
+```C#
+namespace System
+{
+    public sealed class ObsoleteAttribute : Attribute
+    {
+        // Existing:
+        //
+        // public ObsoleteAttribute();
+        // public ObsoleteAttribute(string message);
+        // public ObsoleteAttribute(string message, bool error);
+        // public bool IsError { get; }
+        // public string Message { get; }
+
+        // New:
+        public string DiagnosticId { get; set; }
+        public string UrlFormat { get; set; }
+        public string Url => UrlFormat == null ? null : string.Format(UrlFormat, DiagnosticId);
+    }
+}
+```
+
+Instead of taking the URL directly, the API takes a format string. This allows
+having a generic URL that includes the diagnostic ID. This avoids having to
+repeat the ID twice and thus making a copy & paste mistake.
+
+### Compiler Changes
+
+The compilers would be enlightened about these new properties and use them as
+follows:
+
+* If `DiagnosticId` is `null`, use the existing diagnostic ID (e.g. `CS0618` in
+  C#). Otherwise, use the specified warning ID.
+* If `Url` is not `null`, use it as the diagnostic link when rendering them in
+  the IDE.
+* The compiler should assume that `Url` and `DiagnosticId` are independent
+  features, in other words both can be used, either, and neither.
+
+### Samples
+
+Here is how the attribute would be applied:
+
+```C#
+internal static class Obsolete
+{
+    public const string Url = "http://aka.ms/obsolete/{0}";
+
+    // NOTE: Keep this sorted by the the ID's value,
+    //       not the name of the field!
+    public const string SecureStringId = "BCL0001";
+    public const string NonGenericCollectionsId = "BCL0006";
+}
+
+namespace System.Collections
+{
+    [Obsolete(DiagnosticId = Obsolete.NonGenericCollectionsId, UrlFormat=Obsolete.Url)]
+    public class ArrayList
+    {
+        // ...
+    }
+
+    [Obsolete(DiagnosticId = Obsolete.NonGenericCollectionsId, UrlFormat=Obsolete.Url)]
+    public class Hashtable
+    {
+        // ...
+    }
+}
+
+namespace System.Security
+{
+    [Obsolete(DiagnosticId = Obsolete.SecureStringId, UrlFormat=Obsolete.Url)]
+    public sealed class SecureString
+    {
+        // ...
+    }
+}
+```
+
+***OPEN ISSUE**: We need to decide on guidance to avoid ID conflicts with the
+community. We should talk to the Roslyn analyzer team for best practices. Stake
+in the ground could be that we should a prefix like `BCL` that is owned by a
+single party and responsible to avoid duplicated numbers. Other parties would
+need to choose a different prefix.*
+
+### Possible Alternatives
+
+#### String Encoding
+
+Instead of introducing new APIs, we could invent a textual format that we put
+in the attribute's message property, such as:
+
+```C#
+namespace System.Collections
+{
+    [Obsolete("::BCL0006|http://aka.ms/obsolete/{0}")]
```

Comment:
This would be great if we were just lining up help. But if we are also suppressing warnings on this, I think typos would be too serious to use a pattern based approach.

Reply:
Agreed. I've renamed the section from `Possible Alternatives` to `Rejected Alternatives` to make it clear that I think these are bad ideas.